src/function/timestamp_functions.cpp

```diff
@@ -0,0 +1,157 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// timestamp_functions.cpp
+//
+// Identification: src/function/timestamp_functions.cpp
+//
+// Copyright (c) 2015-2017, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "function/timestamp_functions.h"
+
+#include <date/date.h>
+#include <date/iso_week.h>
+#include <inttypes.h>
+
+#include "common/logger.h"
+#include "type/limits.h"
+#include "type/value_factory.h"
+
+namespace peloton {
+namespace function {
+
+uint64_t TimestampFunctions::DateTrunc(uint32_t date_part_type,
+                                       uint64_t value) {
+  DatePartType date_part =
```

Comment:
You should just be able to `static_cast<uint32_t>` if you change `DatePartType` to an enum class that subclasses `uint32_t`

Reply:
I see. In fact I can directly do `static_cast<DatePartType>` even without subclassing `uint32_t`. Is there any difference here? Also I thought `reinterpret_cast` is faster since it doesn't do type checking? (We already checked the validation of the value in `bind_node_visitor`)

Reply:
Neither `static_cast` nor `reinterpret_cast` perform any runtime checks. In general, `static_cast` is the "safe" cast, whereas `reinterpret_cast` allows casting between incompatible types. Since DateTimePart as a subclass of 32-bit ints, using `static_cast` is clear.

Reply:
I see. Thanks for the explanation!